ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/reaper/ReaperTestCase4.java

```diff
@@ -0,0 +1,93 @@
+package com.hp.mwtests.ts.arjuna.reaper;
+
+import com.arjuna.ats.arjuna.common.Uid;
+import com.arjuna.ats.arjuna.coordinator.Reapable;
+import com.arjuna.ats.arjuna.coordinator.TransactionReaper;
+import org.jboss.byteman.contrib.bmunit.BMScript;
+import org.jboss.byteman.contrib.bmunit.BMUnitConfig;
+import org.jboss.byteman.contrib.bmunit.BMUnitRunner;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.junit.Assert.fail;
+
+@RunWith(BMUnitRunner.class)
+@BMScript("reaper")
+@BMUnitConfig(debug = true)
+public class ReaperTestCase4 extends ReaperTestCaseControl {
+
+    private static int _timeout = 5;
+
+    private Reapable reapable1 = new SimpleReapable();
+    private Reapable reapable2 = new SimpleReapable();
+
+    @Test
+    public void test() throws InterruptedException {
+        enableRendezvous("controlledshutdown", false);
+        TransactionReaper transactionReaper = TransactionReaper.transactionReaper();
+        transactionReaper.insert(reapable1, _timeout);
+        // This second Reapable must not be removed as we should check if TransactionReaper waits for it
+        // even when waitForTransaction is false
+        transactionReaper.insert(reapable2, _timeout);
+
+        AtomicBoolean terminated = new AtomicBoolean(false);
+        new Thread(() -> {
+            // The shutdown is from another thread
+            transactionReaper.terminate(false);
+            synchronized (terminated) {
+                terminated.set(true);
+                terminated.notify();
+            }
+        }).start();
+        // This needs to be after the thread above has started so that the other thread can start a shutdown
+        transactionReaper.remove(reapable1);
+
+        synchronized (terminated) {
+            if (!terminated.get()) {
+                try {
+                    // Give some time for the terminate to finish
+                    long beforeWait = System.currentTimeMillis();
```

Comment:
I think we should calculate the timeout around the: `transactionReaper.terminate(false);` call instead of this wait.

Reply:
Done this as well